### PR TITLE
`withoutActuallyEscaping`

### PR DESCRIPTION
**Pull request by jckarter**
`type(of:)` has behavior whose type isn't directly representable in Swift's type system, since it produces both concrete and existential metatypes. In Swift 3 we put in a parser hack to turn `type(of: <expr>)` into a DynamicTypeExpr, but this effectively made `type(of:)` a reserved name. It's a bit more principled to put `Swift.type(of:)` on the same level as other declarations, even with its special-case type system behavior, and we can do this by special-casing the type system we produce during overload resolution if `Swift.type(of:)` shows up in an overload set. This also lays groundwork for handling other declarations we want to ostensibly behave like normal declarations but with otherwise inexpressible types, viz. `withoutActuallyEscaping` from SE-0110.
rdar://problem/27788962

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@DougGregor or @rudkx, this approach seems to work, though when a type-check fails (such as with an obviously mismatched type like `let foo: Int = type(of: 1)`) I get the generic `(_) -> _ is ambiguous without more context` message instead of a more descriptive error about the exact type mismatch. How would I go about improving the error message? Does this look reasonable?

**Comment by slavapestov** [comment]
Can this approach work for Any also? Not that it's reasonable to define your own type that's named Any, but it bothered me the time it's special and not just a name in the Stdlib module. One day we should get rid of AnyObject also, and a similar approach would help with backward compatibility there.

**Comment by jckarter** [comment]
Making `Any` a normal declaration would be cool too. I think a simpler way to handle that would be to inject a `Builtin` typealias or some other stdlib-only token for the empty protocol composition and just declare `typealias Any = <primitive token for empty composition>`. (Or if you want to be clever math weenie, you could special-case recursive requirements in an `&` type to top out to the empty composition so that `typealias Any = Any & Any` works.)

File: stdlib/public/core/Builtin.swift [hunk 1]
```diff
@@ -584,3 +584,42 @@ internal func _unsafeDowncastToAnyObject(fromAny any: Any) -> AnyObject {
...
+  Builtin.staticReport(_trueAfterDiagnostics(), true._value,
+    ("internal consistency error: 'type(of:)' operation failed to resolve"
+     as StaticString).utf8Start._rawValue)
+  Builtin.unreachable()
+}
+
+/// Return the dynamic type of a value.
+@_transparent
+@_semantics("typechecker.withoutActuallyEscaping")
+public func withoutActuallyEscaping<ClosureType, ResultType>(
```
Comment (rudkx) [inline stdlib/public/core/Builtin.swift]: Did you actually mean to include this since there's no support for the semantic yet?

**Comment by jrose-apple** [comment]
We may still want `Any` for `Any<Collection where Element == Int>`, so I'd be inclined to not touch it for now.

**Comment by rudkx** [comment]
@jckarter Offhand I'm not sure where the best place in CSDiag is to try to generate a better diagnostic for the case you mention above. @xedin has been working in that file a bunch and might have an idea.

File: stdlib/public/core/Builtin.swift (hunk 1 above)
Comment (jckarter) [inline stdlib/public/core/Builtin.swift]: No, I meant to remove that. Thanks!

**Review by rudkx — APPROVED** [review]
BTW, otherwise LGTM!

**Comment by xedin** [comment]
@jckarter Regarding the better error message can you please share an example which returns `ambiguous without more context` because `let foo: Int = type(of: 1)` produces `cannot convert value of type 'Int.Type' to specified type 'Int'` (I tried with your branch as well)...

**Comment by jckarter** [comment]
@xedin You need to pass `-swift-version 4` to get the behavior change—did you pass that? Try the test cases in the PR—they almost all produce the vague diagnostic.

**Comment by xedin** [comment]
@jckarter ah I didn't do that and I see it now. I think the diagnostic changes should go into `FailureDiagnosis::visitApplyExpr` since `type(of:)` is a properly function application now, but I will take a closer look a bit later today and let you know!

**Comment by xedin** [comment]
@jckarter I've figured why there is a regression in the diagnostics - it's related to the fact that `type(of:)` is now expressed as `DynamicTypeOf` constraint classified as `TypeProperty`. 

Because `type(of:)` is a regular `CallExpr` now, it means that `FailureDiagnosis::visitApplyExpr` is going to be used to diagnose it, it's going to attempt to type-check "function" in our case `type(of:)` independently before checking it's arguments. Constraint System produced by `type(of:)` alone consists of a single constraint `T2 dynamicType type of T1` (where type variables form a function type (T1) -> T2) and because `DynamicTypeOf` constraint is neither relational or a member (see https://github.com/apple/swift/blob/master/lib/Sema/CSSolver.cpp#L2324) such system is not going to produce any solutions even with free type variables allowed to be unresolved. Such triggers one more "nested" diagnostics run which can't find any structural or constraint problems because `DynamicTypeOf` constraint is not marked as failed, it's simply unsolved, and falls through to `diagnoseAmbiguity` which produces `ambiguous without more context`.

What I don't yet understand is why wouldn't `type(of: 1)` expression, while being type-checked independently, produce any other constraints besides of `DynamicTypeOf`, I'm going to dig into this further but the good news is once that's fixed diagnostics should "just work".

**Comment by xedin** [comment]
Actually, scratch the last paragraph because DeclRefExpr is type-checked alone there which is never going to produce anything meaningful, so I think the best option would be to try and make it so `type(of:)` produces solutions with unresolved types, if possible...

**Comment by jckarter** [comment]
Thanks for investigating, @xedin! I'll see what I can do with CSDiag.

**Comment by slavapestov** [comment]
Joe, is the swift 4 behavior sufficiently different from swift 3 that it needs to be.behind the flag, or is that temporary?

**Comment by jckarter** [comment]
The only difference would be if someone in Swift 3 mode was trying to overload `type(of:)`, so there's unlikely to be much difference in practice, but better safe than sorry.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@jckarter Assuming this is ready to merge, you might want to update the PR title before doing so.

**Comment by jckarter** [comment]
@rudkx Thanks for the reminder!

**Comment by jtbandes** [comment]
Has any thought been given to naming `withoutActuallyEscaping` with `unsafe` somewhere in it? (Since, isn't the purpose for the user to be able to say "I know better than the compiler in this situation"?) Sorry if this has been discussed already; I didn't see this function mentioned in SE-0110.

**Comment by jrose-apple** [comment]
We're hoping it won't be unsafe because it's something the runtime can verify (using something along the lines of isKnownUniquelyReferenced).

**Pull request merged**